proposed/simple-csharp-programs.md

```diff
@@ -0,0 +1,461 @@
+
+1. Compile and execute in memory
+2. Place artifacts in the temp directory on your machine
+
+There is precedent for (2) in Golang, which means it's probably a viable option. But there is a concern about cleaning the temp folder as well, lest it get too large. If everything is compile and ran in memory, then this isn't a problem. However, given that `#r "nuget"` in F# and .NET Interactive today works by restoring against a project file in the temp folder, the temp folder approach may be the route taken. Other build-time artifacts could be placed there too.
+
+### Multiple files
+
+We will need to decide if Simple C# Programs support more than one file. In other languages, like JavaScript and F#, you can have several loose files in your directory (or other directories) and include them into your program. For example:
+
+**utils.js**
+```js
+const chalk = require('chalk')
+const currentTime = (new Date().toLocaleTimeString([], { hour12: false })).split(':').join('-')
+
+// Pretty print console logs with Chalk
+exports.log = {
+  dim: (text) => console.log(chalk.dim(text)),
+  success: (text) => console.log(chalk.green(text))
+}
+
+// Generates a fileName with current time. e.g. fake-names-19:29:17.json
+exports.getFileName = () => `fake-names-${currentTime}.json`
+```
+
+**program.js**
+
+```js
+const { log, getFileName } = require('../lib/utils')
+
+// use stuff from `utils`
+```
+
+Similarly, in F# there is the `#load "path-to-file.fsx"` directive.
+
+This leads to some questions:
+
+* Are multiple file supported? That is, can I define a `server.cs` file that is my web service, and a `data.cs` that handles some data manipulation that is called from code inside of `server.cs`?
+* If multiple files are supported, what are the guardrails? How can people know if they made a mistake and fix it?
+
+It's also quite likely that some developers will have several programs all as single C# files in the same directory. That means that `dotnet run` would have to understand this and give an appropriate error message about specifying a specific C# file. However, this means inspecting the files to see if they are being used via `#load` directive (or something similar).
+
+Consider the folowing two subfolders.
+
+Directory One:
+
+```
+reboot-server.cs
+check-server-status.cs
+```
+
+Directory Two:
+
+```
+reboot-server.cs
+utils.cs
+```
+
+In Directory Two, the `utils.cs` file is used by `reboot-server.cs`. A developer may expect to be able to just run `dotnet run` in that directory and have the program execute. But if they are in Directory One, they would have to be explicit about which file they are running.
+
+### Coding against a framework alternative syntax
+
+The walkthrough proposes the `#!` directive `#!/path/to/dotnet-aspnet`, but the .NET CLI doesn't have a command to "pull in the ASP.NET Core references", so it wouldn't map to something like that.
+
+An alternative could be to extend `#r` like so:
```

Comment:
This sounds like it's one or the other. Both can be supported at the same time.

Reply:
Yeah, but I don't think we want two ways to do the same thing here.

Reply:
I'm mostly in favor of not having a hashbang approach for this, actually. `#r` is the mini-"language" of referencing things in this kind of context today, and to use ASP.NET you do need to reference the right framework reference. It just happens that the way you express that in a project file is with the SDK attribute.

Reply:
Additionally, the hashbang approach would require a separate executable for every sdk, which isn't good for 3rd-party sdks.

Reply:
I'm also in favor of the `#r` alternative. Maybe the design doc can highlight it is more extensible.